### PR TITLE
fix: extract ingress path to components

### PR DESCRIPTION
**Pull request by wesbillman**

### CODE REVIEW

File: kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt [hunk 1]
```diff
@@ -220,13 +220,24 @@ class SchemaExtractor(
       val pathArg = requireNotNull(argumentLists.second.single().getArgumentExpression()?.text?.trim { it == '\"' }) {
         "$sourcePos Could not extract path from ${verb.name} @Ingress annotation"
       }
+
       MetadataIngress(
-        path = pathArg,
+        path = extractPathComponents(pathArg),
```
Comment (wesbillman) [inline kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt:225]: @worstell Just a heads up. I had to update this bit to support the new path components. Lemme know if you want any changes here! :)

File: backend/schema/normalise.go [hunk 2]
```diff
@@ -69,11 +69,18 @@ func Normalise[T Node](n T) T {
...
 		c.Type = Normalise(c.Type)
 
-	case Decl, Metadata, Type: // Can never occur in reality, but here to satisfy the sum-type check.
+	case *IngressPathLiteral:
+		c.Pos = zero
+
+	case *IngressPathParameter:
+		c.Pos = zero
+
+	case Decl, Metadata, IngressPathComponent, Type: // Can never occur in reality, but here to satisfy the sum-type check.
```
Comment (wesbillman) [inline backend/schema/normalise.go:83]: @alecthomas I sort of guessed on this bit. Does this look ok to ya?

File: backend/schema/validate.go [hunk 3]
```diff
@@ -129,7 +129,8 @@ func ValidateModule(module *Module) error {
 
 		case *Array, *Bool, *DataRef, *Field, *Float, *Int,
 			*Time, *Map, *Module, *Schema, *String, *VerbRef,
-			*MetadataCalls, *MetadataIngress, *Optional:
+			*MetadataCalls, *MetadataIngress, IngressPathComponent,
+			*IngressPathLiteral, *IngressPathParameter, *Optional:
 		case Type, Metadata, Decl: // Union sql.
```
Comment (wesbillman) [inline backend/schema/validate.go:134]: Also guessed here. Lemme know if you want any changes.

File: kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt (hunk 1 above)
Comment (worstell) [reply kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt:225]: lgtm!

File: backend/schema/normalise.go (hunk 2 above)
Comment (alecthomas) [reply backend/schema/normalise.go:83]: Yep correct.

File: go-runtime/compile/schema.go [hunk 4]
```diff
@@ -256,6 +256,21 @@ func visitFuncDecl(pctx *parseContext, node *ast.FuncDecl) (verb *schema.Verb, e
 	return verb, nil
 }
 
+func parsePathComponents(path string) []schema.IngressPathComponent {
```
Comment (alecthomas) [inline go-runtime/compile/schema.go:259]: You'll need to pass the position through to here and add it to the parameters.

**Review by alecthomas — APPROVED** [review]

**Comment by alecthomas** [comment]
This is awesome.

**Pull request merged**